minidump-unwind/src/symbols/debuginfo.rs

```diff
@@ -0,0 +1,488 @@
+        Self::for_module(module)
+    }
+}
+
+struct PerThread<T> {
+    inner: CacheMap<std::thread::ThreadId, UnsafeCell<T>>,
+}
+
+impl<T> Default for PerThread<T> {
+    fn default() -> Self {
+        PerThread {
+            inner: Default::default(),
+        }
+    }
+}
+
+impl<T: Default> PerThread<T> {
+    pub fn with<F, R>(&self, f: F) -> R
+    where
+        F: FnOnce(&mut T) -> R,
+    {
+        // # Safety
+        // We guarantee unique access to the mutable reference because the values are indexed by
+        // thread id: each thread gets its own value which it can freely mutate. We prevent
+        // multiple mutable aliases from being created by requiring a callback function.
+        f(unsafe { &mut *self.inner.cache_default(std::thread::current().id()).get() })
+    }
+}
+
+/// Get the file path with debug information for the given module.
+///
+/// If `unwind_info` is true, returns the path that should contain unwind information.
+fn effective_debug_file(module: &dyn Module, unwind_info: bool) -> PathBuf {
+    // Windows x86_64 always stores the unwind info _only_ in the binary.
+    let ignore_debug_file = unwind_info && cfg!(all(windows, target_arch = "x86_64"));
+
+    let code_file = module.code_file();
+    let code_file_path: &Path = code_file.as_ref().as_ref();
+
+    if !ignore_debug_file {
+        if let Some(file) = module.debug_file() {
+            let file_path: &Path = file.as_ref().as_ref();
+            // Anchor relative paths in the code file parent.
+            if file_path.is_relative() {
+                if let Some(parent) = code_file_path.parent() {
+                    let path = parent.join(file_path);
+                    if path.exists() {
+                        return path;
+                    }
+                }
+            }
+            if file_path.exists() {
+                return file_path.to_owned();
+            }
+        }
+        // else fall back to code file below
+    }
+
+    code_file_path.to_owned()
+}
+
+fn load_unwind_module(module: &dyn Module) -> Option<(Mmap, framehop::Module<ModuleData>)> {
+    let path = effective_debug_file(module, true);
+    let file = match File::open(&path) {
+        Ok(file) => file,
+        Err(e) => {
+            tracing::warn!("failed to open {} for debug info: {e}", path.display());
+            return None;
+        }
+    };
+    // # Safety
+    // The file is presumably read-only (being some binary or debug info file).
+    let mapped = match unsafe { Mmap::map(&file) } {
+        Ok(m) => m,
+        Err(e) => {
+            tracing::error!("failed to map {} for debug info: {e}", path.display());
+            return None;
+        }
+    };
+
+    let objfile = match object::read::File::parse(
+        // # Safety
+        // We broaden the lifetime to static, but ensure that the Mmap which provides the data
+        // outlives all references.
+        unsafe { std::mem::transmute::<_, &'static [u8]>(mapped.as_ref()) },
```

Comment:
I'm not super-fond of this but I also don't see a better way to do it here given we end up with two references